formula/src/main/java/com/instacart/formula/Transition.kt

```diff
@@ -1,57 +1,51 @@
 package com.instacart.formula
 
 /**
- * Defines an intent to transition by emitting a new [State] and optional [Effects].
+ * Transition is a function that is called when an [Event] happens which can produce a [State]
+ * change, executable side-effects or nothing. Take a look at [Result] type.
  */
-sealed class Transition<out State> {
-    companion object {
+fun interface Transition<State, in Event> {
+
+    /**
+     * Defines a transition result which can request a state change and/or some action
+     * to be performed.
+     */
+    sealed class Result<out State> {
```

Comment:
I wonder if `Result` could cause confusion. Would `Kind` be better?

Reply:
Could you explain the confusion with `Result` a bit more? Maybe we can find a different name, though I'm not sure I like `Kind` more. When I think of `Kind` (or `Type`), I usually think of some static enum and not something that contains dynamic data.

Reply:
I thought confusion might come from thinking that a `Result` holds the result of a transition that _has_ occurred, whereas I think this is used to specify the "kind of transition" that _will_ occur.

Reply:
What about `Transition.Event` since it will be sent to `FormulaManager`?

Reply:
>holds the result of a transition that has occurred

My understanding is that it is correct though, Result holds the transformation which has been applied 

We could always use .Data is we wanted something completly neutral, but fwiw Result does not shock me

Reply:
I personally don't have a strong opinion, `Transition.Event` feels a bit better to me.